### PR TITLE
Always return sparse matrices for `spre`, `spost`, and `sprepost`

### PR DESCRIPTION
**Pull request by ytdHuang**
Currently, the return matrix type of `spre` depends on the input, but `spost` and `sprepost` always return sparse matrix.

For superoperators, it should be more feasible to use sparse matrix for large systems. Therefore, I suggest to always return sparse matrix for `spre`.

Also, I find out that if `A` is a sparse matrix, `sparse(A)` will make a copy of it. But we are going to do Kronecker product right away in generating superoperators, I don't think we need to do this copy.

Therefore, I made intrinsic functions for `spre`, `spost`, `sprepost` with multiple dispatches. To deal with different types (dense or sparse) of  input `Opeartor`.

I believe this could improve the performance when creating the Liouvillian.

### CODE REVIEW

**Comment by ytdHuang** [comment]
Seems that `kron` for `Transpose` matrices is not supported before `Julia v1.10`, I will make a multiple dispatch later.

**Review by albertomercurio — COMMENTED** [review]
So the main change is to define `sparse(transpose(sparse(…)))` or `transpose(sparse(…))` depending on the Julia version?

And also convert a dense matrix to a sparse one.

Are these the main changes?

Are this changes compatible with GPU arrays? As far as I remember, `sparse(transpose(sparse(…)))` was needed for GPUs.

File: src/qobj/superoperators.jl [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    _spre(A::AbstractMatrix, Id::AbstractMatrix) = kron(Id, sparse(A))
+    _spre(A::AbstractSparseMatrix, Id::AbstractMatrix) = kron(Id, A)
+    _spost(B::AbstractMatrix, Id::AbstractMatrix) = kron(sparse(transpose(sparse(B))), Id)
+    _spost(B::AbstractSparseMatrix, Id::AbstractMatrix) = kron(sparse(transpose(B)), Id)
+    _sprepost(A::AbstractMatrix, B::AbstractMatrix) = kron(sparse(transpose(sparse(B))), sparse(A))
+    _sprepost(A::AbstractMatrix, B::AbstractSparseMatrix) = kron(sparse(transpose(B)), sparse(A))
+    _sprepost(A::AbstractSparseMatrix, B::AbstractMatrix) = kron(sparse(transpose(sparse(B))), A)
+    _sprepost(A::AbstractSparseMatrix, B::AbstractSparseMatrix) = kron(sparse(transpose(B)), A)
+else
+    _spre(A::AbstractMatrix, Id::AbstractMatrix) = kron(Id, sparse(A))
```
[the remark this answers is not in the document]
Comment (ytdHuang) [reply src/qobj/superoperators.jl:18]: only `_spre` will be the same, I will move it out of the `if-block`

**Comment by ytdHuang** [comment]
> So the main change is to define `sparse(transpose(sparse(…)))` or `transpose(sparse(…))` depending on the Julia version?
> 
> And also convert a dense matrix to a sparse one.
> 
> Are these the main changes?
> 
> Are this changes compatible with GPU arrays? As far as I remember, `sparse(transpose(sparse(…)))` was needed for GPUs.

The main change is to remove unnecessary `sparse()` call. Because this function will make the copy if the input matrix is already a sparse matrix. We only need to transfer the dense matrix to sparse ones.

The `kron` function somehow supports for `Transpose` input type start from `julia 1.10`. 

That is, instead of calling 
```julia
kron(sparse(transpose(sparse(A))), B)
```
like we did before. We can directly use the following call start from julia `v1.10`
```julia
kron(transpose(sparse(A)), B)
```
But we still need to keep the old versions supported, so I made multiple dispatch for different julia versions.

As for the GPU support, I have merged from `main` (to have the CUDA runtests content in #163 ).
I've tested in my local PC, every CUDA extension tests pass.
Since the example in the runtest includes `mesolve`, I think the code here supports GPU computing. 

Maybe they extend some methods in `CUDA.CUSPARSE` so that it is now available.

**Comment by ytdHuang** [comment]
> Ok. Is it also supported for gpus?

I also tried the following cases:
```julia
using CUDA, QuantumToolbox
Xs = cu(sigmax()) # sparse
Xd = cu(sparse_to_dense(sigmax())) # dense

spre(Xs)
spre(Xd)
spost(Xs)
spost(Xd)
```
They all worked, and returns
```
Quantum Object:   type=SuperOperator   dims=[2]   size=(4, 4)
4×4 CUDA.CUSPARSE.CuSparseMatrixCSC{ComplexF64, Int32} with 4 stored entries:
      ⋅            ⋅       1.0 + 0.0im       ⋅
      ⋅            ⋅            ⋅       1.0 + 0.0im
 1.0 + 0.0im       ⋅            ⋅            ⋅
      ⋅       1.0 + 0.0im       ⋅            ⋅
```
in my local PC

**Review by albertomercurio — APPROVED** [review]

**Comment by ytdHuang** [comment]
@albertomercurio 
Some instability issue happened in ODE solvers under macOS.
Don't know why but I think this is something that is beyond our scope. Maybe we can just merge it since the other runtests looks fine.

**Pull request merged**